include/tvm/attrs.h

```diff
@@ -0,0 +1,592 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file tvm/attrs.h
+ * \brief TVM attribute module
+ *
+ *  This module enables declaration named attributes
+ *  and support default value setup and bound checking.
+ *
+ * \code
+ *   struct MyAttrs : public tvm::AttrsNode<MyAttrs> {
+ *     float learning_rate;
+ *     int num_hidden;
+ *     std::string name;
+ *     // declare attribute fields in header file
+ *     TVM_DECLARE_ATTRS(MyAttrs, "attrs.MyAttrs") {
+ *       TVM_ATTR_FIELD(num_hidden).set_lower_bound(1);
+ *       TVM_ATTR_FIELD(learning_rate).set_default(0.01f);
+ *       TVM_ATTR_FIELD(name).set_default("hello");
+ *     }
+ *   };
+ *   // register it in cc file
+ *   TVM_REGISTER_NODE_TYPE(MyAttrs);
+ * \endcode
+ *
+ * \sa AttrsNode, TVM_DECLARE_ATTRS, TVM_ATTR_FIELD
+ */
+#ifndef TVM_ATTRS_H_
+#define TVM_ATTRS_H_
+
+#include <unordered_map>
+#include <vector>
+#include <type_traits>
+#include <string>
+#include "./ir.h"
+#include "./base.h"
+#include "./packed_func_ext.h"
+
+namespace tvm {
+/*!
+ * \brief Declare an attribute function.
+ * \param ClassName The name of the class.
+ * \param TypeKey The type key to be used in the tvm node system.
+ */
+#define TVM_DECLARE_ATTRS(ClassName, TypeKey)                   \
+  static constexpr const char* _type_key = TypeKey;             \
+  TVM_DECLARE_NODE_TYPE_INFO(ClassName, ::tvm::BaseAttrsNode);  \
+  template<typename FVisit>                                     \
+  void __VisitAttrs__(FVisit& __fvisit__)  // NOLINT(*)
+
+
+/*!
+ * \brief Declare an attribute field.
+ * \param FieldName The field name.
+ */
+#define TVM_ATTR_FIELD(FieldName) \
+  __fvisit__(#FieldName, &FieldName)
+
+
+/*! \brief Error throwed during attribute checking */
+struct AttrError : public dmlc::Error {
+  /*!
+   * \brief constructor
+   * \param msg error message
+   */
+  explicit AttrError(const std::string &msg)
+      : dmlc::Error(msg) {}
+};
+
+/*!
+ * \brief Information about attribute fields in string representations.
+ */
+struct AttrFieldInfo {
+  /*! \brief name of the field */
+  std::string name;
+  /*! \brief type of the field in string format */
+  std::string type_info;
```

Comment:
Why do we store the type in string format? is this the string key?

Reply:
This is used in docstring, clarified in the comment